### PR TITLE
Add bert ci test

### PR DESCRIPTION
**Pull request by jh619lee**
## 문제상황
- mlperf_submission bert 모델과 huggingface bert 모델 간의 동치 확인(bit exactly same) 을 위한 테스트 구현 필요
- 여러 버전의 calib dataloader 관리 및 importing이 용이하도록 업데이트 필요.

## 목적(해결방향)
- 동치 확인을 위한 ci test  추가
- calib dataloader i/o 정리

## 구현 설명 Abstract
- mlperf loadgen 시나리오에 따라 batch_size=1인 경우에 대하여 동치 확인을 수행하는 ci test 추가.
- calib dataloader와 squad data loading 부분을 분리하여 dataloader의 i/o 통일

## 구현 설명 Detail (ex. 추가된 argument)
- run_ci.py 추가
   - 동작 순서 
       - huggingface bert 모델과 mlpefr submission에 대한 calibration을 각각 수행
       - qparam 비교 -> qparam 동치가 보장되지 않을 경우 error raise
       - qlv4 모델 변환 후, batch =1 에 대한 model forwarding 수행
       - logits 비교 -> valid sequence에 대하여 bit exactly same 비교 수행하고, 동치가 보장되지 않을 경우  error raise
- calib_dataloader.py 업데이트
    - `calib_eval_features=load_bert_calibration_data(qsl, n_calib)`를 정의하여, squad data loading 부분 분리
    - 모든 calib_dataloader의 arg인자 통일 
        - args: calib_eval_features , batch_size, n_calib
            - calib_eval_feature: list 타입의 data로, 각  data는 input_ids, segment_ids, input_mask를 attr로 가짐.
        - kwargs: dataloader 종류에 따라 다르게 정의되는 변수.
            -ex) pad_token_id, bucket_size, include_position_ids 등
            
## test 통과 내역 (CI 통과내역과 어떤 machine (GPU pod or NPU machien or local)에서 테스트했는지)
- [ ] local machine with 3090
- [ ] GPU pod
- [ ] warboy pod
  - `"python main.py --scenario Offline --model-path ./model/ --dataset-path ./data/cnn_eval_accuracy_ci.json --model_script_path ./quantization/model_script/Qlevel4_RGDA0-W8A8KV8-PTQ-SMQ.yaml --gpu --accuracy --use_mcp --calib-dataset-path ./data/cnn_dailymail_calibration.json --recalibrate --model_source [transformers or furiosa_llm_original]"

## TODO (ex. 후속PR예고)
-

### CODE REVIEW

**Comment by jeongin-yun** [comment]
run_ci.py 실행 script 도 부탁드립니다.

**Comment by jh619lee** [comment]
@jeongin-yun 정인님, bash 형태 script 말씀하시는 걸까요?

**Comment by jeongin-yun** [comment]
@jh619lee python 실행 script 면 됩니다.

**Comment by jh619lee** [comment]
일단 현재 필요한 args는 코드 상에서 설정되고 있어서 `python run_ci.py`로만 실행하면 됩니다. 추후 compact_mask 모델도 같은 코드로 테스트할 경우에는 script arg가 필요할 것 같은데, compact_mask 모델 확인 후 추가 작업 진행하겠습니다.

File: language/bert/quantization/calib_dataloader/calib_dataloader.py [hunk 1]
```diff
@@ -1,28 +1,38 @@
...
+        (
+            input_ids,
+            token_type_ids,
+            attention_mask,
+            position_ids,
+            packed_target_locations,
+        ) = greedy_attention_packing_bert(
+            input_ids=torch.LongTensor(feature.input_ids).unsqueeze(0),
+            token_type_ids=torch.LongTensor(feature.segment_ids).unsqueeze(0),
+            bucketized_attention_mask=torch.LongTensor(feature.input_mask).unsqueeze(0),
```
Comment (jeongin-yun) [inline language/bert/quantization/calib_dataloader/calib_dataloader.py]: compact causcal mask 사용시 필요한 처리 인가요?

Comment (jh619lee) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py]: compact_causal_mask와 별개로 놓쳤던 부분이라 수정하였습니다.

Comment (jeongin-yun) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py]: 위 부분은 generator 내부 동작이랑 약간 다른거 같은데 왜 그런가요?
(bucket_pad 사용하지 않음)

Comment (jh619lee) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py]: bucket_size를 max_seq_length로 지정해준 상황이라 pad가 동작하지 않아서 제외 시켰는데, 파악하기 쉽도록 형태를 일치시키는 commit 생성하였습니다.

**Comment by jh619lee** [comment]
generator 수정된 부분 반영한 업데이트도 본 PR에서 하겠습니다!

File: language/bert/quantization/calib_dataloader/calib_dataloader.py [hunk 2]
```diff
@@ -1,69 +1,122 @@
...
-                input_ids=bucket_pad(batch["input_ids"]),
-                token_type_ids=bucket_pad(batch["token_type_ids"]),
-                bucketized_attention_mask=bucket_pad(batch["attention_mask"]),
-                pad_token_id=pad_token_id,
-                compact_mask=False,
-            )
+    for feature in calib_eval_features:
+        input_ids = torch.LongTensor(feature.input_ids).unsqueeze(0).to(self.dev)
+        attention_mask = torch.LongTensor(feature.input_mask).unsqueeze(0).to(self.dev)
+        token_type_ids = torch.LongTensor(feature.segment_ids).unsqueeze(0).to(self.dev)
```
Comment (jeongin-yun) [inline language/bert/quantization/calib_dataloader/calib_dataloader.py]: function 인데, self.dev 가 존재하나요?

Comment (jh619lee) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py]: 수정 commit을 함께 올릴 예정입니다!

File: language/bert/quantization/calib_dataloader/calib_dataloader.py [hunk 3]
```diff
@@ -1,69 +1,122 @@
...
 
-    data_list = []
     if n_calib != -1:
         calib_eval_features = calib_eval_features[0:n_calib]
-    for feature in calib_eval_features:
-        data_list.append({
-            'input_ids': torch.LongTensor(feature.input_ids),
-            'attention_mask': torch.LongTensor(feature.input_mask),
-            'token_type_ids': torch.LongTensor(feature.segment_ids),
-        })
```
Comment (jeongin-yun) [inline language/bert/quantization/calib_dataloader/calib_dataloader.py:65]: 이 부분은 mlperf data dependant 한 처리여서, 여기 남아 있으면, make_packed_calib_data_loader 를 mcp ci 나 추후에 furiosa-llm-models generator 에 immigration 하기 쉬울거 같습니다.

현 PR에선 제외하고, 후속 작업으로 고려하겠습니다.

Comment (jh619lee) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py:65]: model input data가 input_ids, attention_mask, token_type_ids를 포함해야하기 때문에 , 하나의 feature가 해당 정보를 모두 포함한다는 가정이 make_dataloder에서도 자연스럽다고 생각했는데, attr로 접근해야하는 방식이 혹시 부자연스럽게 보이시는 걸까요?

**Comment by jh619lee** [comment]
generator 수정 반영 후, hf bert와 mlperf_submission bert 동치 확인되었습니다. `python run_ci.py`로 테스트 가능합니다.

File: language/bert/quantization/calib_dataloader/calib_dataloader.py (hunk 3 above)
Comment (jeongin-yun) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py:65]: 'input_mask', 'segment_ids' 와 같은 attr name 이 해당 data specific 한 정보라서, make_dataloder 를 다른 곳에서 그대로 사용 불가능합니다.

**Review by jeongin-yun — APPROVED** [review]
LGTM

File: language/bert/quantization/calib_dataloader/calib_dataloader.py (hunk 3 above)
Comment (jh619lee) [reply language/bert/quantization/calib_dataloader/calib_dataloader.py:65]: 해당 이슈는 다음 PR에서 작업하겠습니다!
